### ISSUE TITLE
RunUMAP segfault address 0x3d8, cause 'memory not mapped'

### ISSUE THREAD

**Issue by petukhovaaa** [satijalab/seurat#10384]
Hi everyone!

First of all, thank you for the Seurat package! I'm currently analyzing a big scRNA-seq dataset (~500 thousand cells) and recently I've started to get the following issue while running RunUMAP:

```
02:26:40 UMAP embedding parameters a = 0.9922 b = 1.112
02:26:40 Read 496899 rows and found 20 numeric columns
02:26:40 Using Annoy for neighbor search, n_neighbors = 30
02:26:40 Building Annoy index with metric = cosine, n_trees = 50
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
02:27:43 Writing NN index file to temp file /scratch/vslab/tmp/RtmpaHeWzx/file185a00267cad84
02:27:43 Searching Annoy index using 1 thread, search_k = 3000
02:32:04 Annoy recall = 100%
02:32:05 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
02:32:27 Initializing from normalized Laplacian + noise (using RSpectra)

 *** caught segfault ***
address 0x3d8, cause 'memory not mapped'

Traceback:
 1: .Primitive(".Call")("eigs_shift_sym", new("dgCMatrix", i = c(0L, 2677L, 25660L, 32568L, 34250L, 49410L, 78379L, 80834L, 93465L, 99294L, 106465L, 107561L, 112738L, 116099L, 119913L, 126033L, 137201L, 142772L, 142863L, 145963L, 146010L, 148230L, 149267L, 149553L, 151028L, 267998L, 317184L, 324964L, 334559L, 353419L, 1L, 280L, 1284L, 1857L, 2044L, 2330L, 3666L, 3850L, 4681L, 4870L, 5622L, 8999L, 9452L, 11414L, 12211L, 13806L, 15476L, 15513L, 15690L, 16559L, 17111L, 17565L, 17913L, 22181L, 43582L, 45475L, 47595L, 48946L, 51080L, 53767L, 55279L, ***** a lot of numbers ***** 1, -0.00161622044123257, -0.0743250394066115, -0.00959966651761123, -0.0149240923653673, -0.00106435669413037, -0.0106235708493492, -0.00616863150804804, -0.0623613390107527, -0.00898284813455874, -0.00572485050887022, -0.0156445599590061, -0.116141422029551, -0.073835109009995, -0.00689690224882944, -0.0517252104039432, -0.0181959486605138, -0.00812766839374274, -0.00992588374971378, -0.00346979232260604, -0.0759918370881799, -0.0127794668492782, -0.00378824013584267, -0.0213076459357248, -0.00723175525419568, -0.00837845361418333, -0.00404426829336243, -0.0984921910529418, -0.106234571545916, -0.0130846692802979, -0.00234695577953803, -0.0111868252413289, -0.00255893486405049, -0.00817168602110821, -0.00601435689054214, -0.00666314848529302, -0.0230461185380525, -0.0054263176979031, -0.0110256728641533, -0.00764566045597079, -0.00212696360534456, -0.0094307742804036, -0.00416761897170264, -0.00852565886404069, -0.00276254064867692, -0.00187190431052392, -0.0146556202224824, -0.00491490813779362, -0.0320797934063784, -0.0492956873887806, 1), factors = list()), 496899L, 3L, list(which = c(LM = 0L),     ncv = 20, tol = 1e-04, maxitr = 1000, retvec = TRUE, user_initvec = FALSE,     sigma = 1e-09, use_lower = TRUE), 6L, PACKAGE = "RSpectra")                                                              
 2: do.call(.Call, args = dot_call_args)
 3: eigs_real_sym(A, nrow(A), k, which, sigma, opts, mattype = "sym_dgCMatrix",     extra_args = list(use_lower = as.logical(lower)))
 4: eigs_sym.dgCMatrix(L, k = k, which = "LM", sigma = 1e-09, opts = opt)
 5: RSpectra::eigs_sym(L, k = k, which = "LM", sigma = 1e-09, opts = opt)
 6: doTryCatch(return(expr), name, parentenv, handler)
 7: tryCatchOne(expr, names, parentenv, handlers[[1L]])
 8: tryCatchList(expr, classes, parentenv, handlers)
 9: tryCatch(RSpectra::eigs_sym(L, k = k, which = "LM", sigma = 1e-09,     opts = opt), error = function(c) {    tsmessage("RSpectra shifted calculation also failed")    NULL})
10: withCallingHandlers(expr, warning = function(w) if (inherits(w,     classes)) tryInvokeRestart("muffleWarning"))
11: suppressWarnings(res <- tryCatch(RSpectra::eigs_sym(L, k = k,     which = "LM", sigma = 1e-09, opts = opt), error = function(c) {    tsmessage("RSpectra shifted calculation also failed")    NULL}))
12: rspectra_eigs_sym(L, ndim, verbose = verbose)
13: rspectra_normalized_laplacian_init(A, ndim, verbose = FALSE)
14: spectral_init(V, ndim = n_components, verbose = verbose)
15: uwot(X = X, n_neighbors = n_neighbors, n_components = n_components,     metric = metric, n_epochs = n_epochs, alpha = learning_rate,     scale = scale, init = init, init_sdev = init_sdev, spread = spread,     min_dist = min_dist, set_op_mix_ratio = set_op_mix_ratio,     local_connectivity = local_connectivity, bandwidth = bandwidth,     gamma = repulsion_strength, negative_sample_rate = negative_sample_rate,     a = a, b = b, nn_method = nn_method, n_trees = n_trees, search_k = search_k,     method = "umap", approx_pow = approx_pow, n_threads = n_threads,     n_sgd_threads = n_sgd_threads, grain_size = grain_size, y = y,     target_n_neighbors = target_n_neighbors, target_weight = target_weight,     target_metric = target_metric, pca = pca, pca_center = pca_center,     pca_method = pca_method, pcg_rand = pcg_rand, fast_sgd = fast_sgd,     ret_model = ret_model || "model" %in% ret_extra, ret_nn = ret_nn ||         "nn" %in% ret_extra, ret_fgraph = "fgraph" %in% ret_extra,     ret_sigma = "sigma" %in% ret_extra, ret_localr = "localr" %in%         ret_extra, batch = batch, opt_args = opt_args, epoch_callback = epoch_callback,     binary_edge_weights = binary_edge_weights, tmpdir = tempdir(),     verbose = verbose, dens_scale = dens_scale, seed = seed,     nn_args = nn_args, rng_type = rng_type)
16: umap(X = object, n_threads = nbrOfWorkers(), n_neighbors = as.integer(x = n.neighbors),     n_components = as.integer(x = n.components), metric = metric,     n_epochs = n.epochs, learning_rate = learning.rate, min_dist = min.dist,     spread = spread, set_op_mix_ratio = set.op.mix.ratio, local_connectivity = local.connectivity,     repulsion_strength = repulsion.strength, negative_sample_rate = negative.sample.rate,     a = a, b = b, fast_sgd = uwot.sgd, verbose = verbose, ret_model = return.model)
17: RunUMAP.default(object = data.use, reduction.model = reduction.model,     return.model = return.model, assay = assay, umap.method = umap.method,     n.neighbors = n.neighbors, n.components = n.components, metric = metric,     n.epochs = n.epochs, learning.rate = learning.rate, min.dist = min.dist,     spread = spread, set.op.mix.ratio = set.op.mix.ratio, local.c
onnectivity = local.connectivity,     repulsion.strength = repulsion.strength, negative.sample.rate = negative.sample.rate,     a = a, b = b, uwot.sgd = uwot.sgd, seed.use = seed.use, 
metric.kwds = metric.kwds,     angular.rp.forest = angular.rp.forest, densmap = densmap,     dens.lambda = dens.lambda, dens.frac = dens.frac, dens.var.shift = dens.var.shift,     redu
ction.key = reduction.key %||% Key(object = reduction.name,         quiet = TRUE), verbose = verbose)
18: RunUMAP(object = data.use, reduction.model = reduction.model,     return.model = return.model, assay = assay, umap.method = umap.method,     n.neighbors = n.neighbors, n.components
 = n.components, metric = metric,     n.epochs = n.epochs, learning.rate = learning.rate, min.dist = min.dist,     spread = spread, set.op.mix.ratio = set.op.mix.ratio, local.connectiv
ity = local.connectivity,     repulsion.strength = repulsion.strength, negative.sample.rate = negative.sample.rate,     a = a, b = b, uwot.sgd = uwot.sgd, seed.use = seed.use, metric.k
wds = metric.kwds,     angular.rp.forest = angular.rp.forest, densmap = densmap,     dens.lambda = dens.lambda, dens.frac = dens.frac, dens.var.shift = dens.var.shift,     reduction.ke
y = reduction.key %||% Key(object = reduction.name,         quiet = TRUE), verbose = verbose)
19: RunUMAP.Seurat(seurat_mat, dims = 1:pc_num)
20: RunUMAP(seurat_mat, dims = 1:pc_num)
An irrecoverable exception occurred. R is aborting now ...
/var/spool/slurmd/job39761828/slurm_script: line 45: 1595904 Segmentation fault      (core dumped) Rscript data_analysis_crash.R $path $project $input $output
```

I run the standard Seurat workflow as a Slurm-job on a server with 700 Gb of memory, so I don't think that the memory is the issue. Also, this code worked well on this or bigger datasets, but now something goes wrong. And I already saw some issues where the reason was in uwot, RSpectra and Rcpp* packages versions, but rebuilding my conda environment didn't help me yet. This is my session info at the moment:

```
R version 4.3.3 (2024-02-29)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: Rocky Linux 9.7 (Blue Onyx)

Matrix products: default
BLAS/LAPACK: /ref/vslab/software/conda/envs/r-env_copy/lib/libopenblasp-r0.3.32.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: America/Chicago
tzcode source: system (glibc)

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] dendsort_0.3.4     dendextend_1.19.1  ape_5.8-1          lubridate_1.9.4   
 [5] forcats_1.0.0      stringr_1.5.2      purrr_1.1.0        readr_2.1.5       
 [9] tidyr_1.3.1        tidyverse_2.0.0    Signac_1.14.0      hdf5r_1.3.12      
[13] ggplot2_3.5.2      tibble_3.3.0       gridExtra_2.3      dplyr_1.1.4       
[17] Seurat_5.3.0       SeuratObject_5.2.0 sp_2.2-0           devtools_2.4.5    
[21] usethis_3.2.1      Matrix_1.6-5      

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      jsonlite_2.0.0          magrittr_2.0.3         
  [4] spatstat.utils_3.1-5    farver_2.1.2            zlibbioc_1.48.0        
  [7] fs_1.6.6                vctrs_0.6.5             ROCR_1.0-11            
 [10] Rsamtools_2.18.0        memoise_2.0.1           spatstat.explore_3.5-2 
 [13] RCurl_1.98-1.17         RcppRoll_0.3.1          htmltools_0.5.8.1      
 [16] sctransform_0.4.2       parallelly_1.45.1       KernSmooth_2.23-26     
 [19] htmlwidgets_1.6.4       ica_1.0-3               plyr_1.8.9             
 [22] plotly_4.11.0           zoo_1.8-14              cachem_1.1.0           
 [25] igraph_2.1.4            mime_0.13               lifecycle_1.0.4        
 [28] pkgconfig_2.0.3         R6_2.6.1                fastmap_1.2.0          
 [31] GenomeInfoDbData_1.2.11 fitdistrplus_1.2-4      future_1.67.0          
 [34] shiny_1.11.1            digest_0.6.37           colorspace_2.1-1       
 [37] S4Vectors_0.40.2        patchwork_1.3.2         tensor_1.5.1           
 [40] RSpectra_0.16-1         irlba_2.3.5.1           pkgload_1.4.0          
 [43] GenomicRanges_1.54.1    progressr_0.15.1        timechange_0.3.0       
 [46] spatstat.sparse_3.1-0   httr_1.4.7              polyclip_1.10-7        
 [49] abind_1.4-5             compiler_4.3.3          remotes_2.5.0          
 [52] bit64_4.6.0-1           withr_3.0.2             BiocParallel_1.36.0    
 [55] viridis_0.6.5           fastDummies_1.7.5       pkgbuild_1.4.8         
 [58] MASS_7.3-60.0.1         sessioninfo_1.2.3       tools_4.3.3            
 [61] lmtest_0.9-40           httpuv_1.6.16           future.apply_1.20.0    
 [64] goftest_1.2-3           glue_1.8.0              nlme_3.1-168           
 [67] promises_1.3.3          Rtsne_0.17              cluster_2.1.8.1        
 [70] reshape2_1.4.4          generics_0.1.4          gtable_0.3.6           
 [73] spatstat.data_3.1-8     tzdb_0.5.0              hms_1.1.3              
 [76] data.table_1.17.8       XVector_0.42.0          BiocGenerics_0.48.1    
 [79] spatstat.geom_3.5-0     RcppAnnoy_0.0.22        ggrepel_0.9.6          
 [82] RANN_2.6.2              pillar_1.11.0           spam_2.11-1            
 [85] RcppHNSW_0.6.0          later_1.4.4             splines_4.3.3          
 [88] lattice_0.22-7          survival_3.8-3          bit_4.6.0              
 [91] deldir_2.0-4            tidyselect_1.2.1        Biostrings_2.70.1      
 [94] miniUI_0.1.2            pbapply_1.7-4           IRanges_2.36.0         
 [97] scattermore_1.2         stats4_4.3.3            matrixStats_1.5.0      
[100] stringi_1.8.7           lazyeval_0.2.2          codetools_0.2-20       
[103] cli_3.6.5               uwot_0.2.3              xtable_1.8-4           
[106] reticulate_1.43.0       Rcpp_1.1.0              GenomeInfoDb_1.38.1    
[109] globals_0.18.0          spatstat.random_3.4-1   png_0.1-8              
[112] spatstat.univar_3.1-4   parallel_4.3.3          ellipsis_0.3.2         
[115] dotCall64_1.2           profvis_0.4.0           urlchecker_1.0.1       
[118] bitops_1.0-9            listenv_0.9.1           viridisLite_0.4.2      
[121] scales_1.4.0            ggridges_0.5.7          crayon_1.5.3           
[124] rlang_1.1.6             fastmatch_1.1-6         cowplot_1.2.0       
```

Will really appreciate any help and any insights related to this issue! Thank you!

**Comment by anashen** [comment]
Hi @petukhovaaa, the problem seems to occur during `uwot`’s spectral initialization step (`Initializing from normalized Laplacian + noise (using RSpectra)` is the last thing you see before the memory error). This can hang or fail to converge when the nearest-neighbor graph has multiple disconnected components. `uwot` documents this behavior and describes alternative initialization methods here: https://jlmelville.github.io/uwot/articles/init.html.

The behavior of  `RunUMAP()` stalling at the `RSpectra` initialization step was also reported in an [earlier Seurat issue](https://github.com/satijalab/seurat/issues/9352). 

As detailed in this issue, you can try approximate global spectral initialization - we added a new parameter `uwot.init` to `RunUMAP` that allows for customizing UMAP initialization, so you can pass "agspectral" to this. You will first need to install the development version of Seurat (`devtools::install_github("satijalab/seurat", ref = "main")`) and then restart your R session.

You can also try the sketch workflow or adjusting the arguments to `RunUMAP`, but I think this problem can be resolved by trying a different initialization method. Please let me know if you have any issues or further questions. Thanks!

**Comment by anashen** [comment]
Closing for now - please see comment above, and as an update, the `uwot.init` parameter was made available in official releases starting with v5.5.1.

Feel free to reopen with any further issues or questions. Thanks!
